raft/server.cc

```diff
@@ -366,12 +366,23 @@ future<> server_impl::wait_for_entry(entry_id eid, wait_type type, seastar::abor
 
             _stats.waiters_awoken++;
 
-            if (term && *term == eid.term) {
-                co_return;
+            if (!term) {
+                throw commit_status_unknown();
+            }
+
+            if (*term != eid.term) {
+                throw dropped_entry();
             }
-            // An entry with a different term has been committed.
-            // It implies this entry was dropped.
-            throw dropped_entry();
+
+            if (type == wait_type::applied && _fsm->log_last_snapshot_idx() >= eid.idx) {
+                // We know the entry was committed but the wait type is `applied`
+                // and we don't know if the entry was applied with `state_machine::apply`
+                // (we may've loaded a snapshot before we managed to apply the entry).
+                // As specified by `add_entry`, throw `commit_status_unknown` in this case.
+                throw commit_status_unknown();
```

Comment:
Could it happen that the entry was applied locally (not via snapshot), but the leader created a snapshot over it, and so we will return commit_status_unknown. If snapshots are created fast enough, that can make it a liveness problem.

Reply:
The scenario which this patch addresses is very unlikely, and your scenario is even more unlikely. Probably impossible in practice.

The following would have to happen:
1. We forward the entry to the leader (in `server_impl::add_entry` we call `_rpc->send_add_entry(leader, command)`)
2. The leader appends the entry and sends a response to `send_add_entry`, let's call this message `M`
3. The leader does a round-trip to a quorum of voters and commits the entry, sends the updated log and commit index to us
4. The leader creates a snapshot and sends it to us
5. We apply the entry using the updated commit index, then we apply the snapshot
6. Only now `M` arrives to us, so `send_add_entry` finishes and we enter `wait_for_entry`

And this could only happen at most once in 500 entries, assuming the default snapshotting frequency.

In fact from the description above it seems that message reordering is required for even the scenario addressed by this patch to happen. Reordering is possible (and happens often) in my nemesis test but it probably won't be possible in production (if we use TCP) or at least very unlikely.

Reply:
wait_for_entry() is also called when not forwarding, and reordering can happen in the CPU scheduler.

Reply:
Then the scenario is even less likely, the window in which commiting the entry and creating the snapshot must happen before we enter `wait_for_entry` is even shorter, much much shorter. In the forwarding scenario it would have to happen when the reply to `send_add_entry` was in transit, in non-forwarding it would have to happen between `co_await execute_add_entry` and `co_await wait_for_entry` (admittedly the snapshot does not need to be sent over the network in this case, but there is still the round-trip necessary to commit the entry).

Reply:
Then the scenario is even less likely, the window in which commiting the entry and creating the snapshot must happen before we enter `wait_for_entry` is even shorter, much much shorter. In the forwarding scenario it would have to happen when the reply to `send_add_entry` was in transit, in non-forwarding it would have to happen between `co_await execute_add_entry` and `co_await wait_for_entry` (admittedly the snapshot does not need to be sent over the network in this case, but there is still the round-trip necessary to commit the entry).

Reply:
Fair enough. Although, I hate relying on likelihood, because it may change the future, and it's hard to reason about (space of potential executions is large).

An explicit solution would be cleaner, e.g. have the local apply notify the waiter, not rely on indexes.

Reply:
Well TBH when we talk about liveness, likelihood is the only thing we can rely on (FLP theorem).

Reply:
The likelihood of the system behaving well increases the less we depend on likelihood.